# 내 프로젝트 v1 - 배터리 셀 검사 (견본)

> **v1 = 첫 번째 판.** 완성본이 아니라 오늘 만드는 첫 판이라는 뜻입니다.
>
> ⭐ **실습 16 · 17 · 18 · 19를 한 노트북에 담은 견본입니다.** 오늘 네 칸이 전부 여기 들어 있어요.
>
> **이것은 견본입니다.** 화면으로 함께 만든 것을 그대로 담았습니다.
> 여러분은 **본인 데이터로** 같은 순서를 한 벌 만드시면 됩니다. 이 파일은 고치지 말고 열어보기만 하세요.

| 실습 | 이 노트북의 어디 |
|---|---|
| 실습 16 - 문제 정의서 | 문제 정의 · 쓰는 데이터 · 한계 |
| 실습 17 - 모델링 계획표 | 모델링 계획 |
| 실습 18 - 모델링 v1 | 문법 노트 · 내 갈래 · 코드 · 받은 코드 확인 |
| 실습 19 - 결과 한 장 | v1 결과 · 다음에 할 일 |

> 실습마다 맨 끝에 **도전** 칸이 하나씩 붙어 있습니다. 도전은 안 하셔도 되지만, **네 개 다 있으면 포트폴리오에 그대로 실립니다.**


# 실습 16 - 문제 정의서

## 문제 정의

**한 문장** : 배터리 셀 검사 측정값으로, 최종 검사 전에 불합격 가능성이 높은 셀을 먼저 골라낸다

| 물음 | 내 답 |
|---|---|
| 누가 쓰나 | 셀 검사 공정 담당자 |
| 무엇을 결정하나 | 이 셀을 정밀 재검사로 보낼지 말지 |
| 언제 결정하나 | 자동 측정이 끝나고 값이 표에 올라온 직후 |
| 이 판단이 늦으면 | 불합격 셀이 조립 공정까지 넘어가고, 완제품에서 발견되면 묶음 전체를 다시 봐야 한다 |

## 쓰는 데이터

| 항목 | 내용 |
|---|---|
| 출처 | 수업에서 받은 셀 검사 기록 |
| 규모 | 2847행 x 14열 |
| 입력으로 쓸 열 | 숫자로 된 측정 열 8개 |
| 결과 열 (TARGET) | result |
| 결과 열의 값 | 합격 / 불합격 |
| 기준선이 될 숫자 | 불합격 133건 (4.67%) |

## 한계

- 라인과 근무조가 표에 있지만 첫 판에서는 숫자 열만 썼다. 조별 차이는 아직 보지 않았다
- 기록 기간이 짧아 계절이나 원료 교체 같은 긴 흐름은 담기지 않았다
- 특정 측정값이 높은 구간에서 불합격이 많이 관찰되더라도, 그것이 원인이라고는 말할 수 없다

---

### 도전 - 말로 하는 30초 버전

배터리 셀은 검사에서 측정값이 열 개 넘게 나옵니다. 사람이 다 보기는 어려워서, 측정값만으로 불합격 가능성이 높은 셀을 먼저 골라내는 것을 만들었습니다. 셀 검사 담당자가 정밀 재검사로 보낼 셀을 고를 때 쓰는 것을 생각하고 만들었고, 전체의 4.67%만 불합격인 데이터라 정확도만 보면 안 되는 문제였습니다. 그래서 지목 건수와 그중 진짜였던 건수를 함께 봤습니다.

# 실습 17 - 모델링 계획표

## 모델링 계획

| # | 하는 일 | 내 경우엔 | 막힐 것 같나 |
|---|---|---|---|
| 1 | 파일 불러오기 | ../../data/day04_battery.csv | |
| 2 | 빈칸 채우기 | 숫자 열의 빈칸을 그 열의 중앙값으로 | |
| 3 | 결과 열을 숫자로 | result가 "불합격"이면 1, 아니면 0 | |
| 4 | 입력과 정답 나누기 | 입력은 측정 열 8개(features), 정답은 label | |
| 5 | 학습용·시험용 나누기 | 8대 2, 비율을 양쪽에 맞춰서 | |
| 6 | 기준 모델 점수 재기 | 전부 합격이라 답하는 모델 | |
| 7 | 모델 하나 학습시키기 | 표준화한 뒤 로지스틱 회귀 | 막힐 것 같음 |
| 8 | 결과 읽기 | 기준 모델과 나란히 놓고 비교 | 막힐 것 같음 |

### 이번 v1에서 볼 숫자

- 볼 숫자 : 기준 모델 정확도 / 내 모델 정확도 / 지목 건수 / 그중 진짜 / 정밀도 / 재현율
- 이걸 보는 이유 : 불합격이 4.67%라 전부 합격이라 답해도 95%가 넘는다. 정확도만 보면 잘한 걸로 착각한다

---

### 도전 - 입력에서 빼야 할 열이 있나 (데이터 누수 점검)

| 종류 | 있나 | 열 이름 | 어떻게 할까 |
|---|---|---|---|
| 번호·순번 열 | 있음 | cell_id | 글자로 되어 있어 숫자 열에 안 잡힌다. 그대로 두면 자동으로 빠진다 |
| 결과 뒤에 생기는 열 | 없음 | - | - |
| 결과 열의 사본 | 없음 | - | - |

- 확인한 것 : 숫자 열 8개는 모두 검사 장비에서 나온 측정값이고, 합격·불합격이 정해지기 전에 찍힌다
- 결론 : 입력 8개를 그대로 쓴다(DROP은 빈 목록). 다만 정확도가 99%를 넘으면 이 표를 다시 본다


# 실습 18 - 모델링 v1

## 모델링 v1

### 문법 노트 - 코드에서 찾을 이름

⭐ **오늘부터 변수 이름을 영어로 씁니다.** 이 코드가 그대로 GitHub 저장소에 올라가고 면접에서 열리기 때문입니다. 현업에서 쓰는 이름 그대로 익혀두면 남의 코드를 읽을 때도 바로 붙습니다.

| 이름 | 뜻 | 무엇이 들어 있나 |
|---|---|---|
| DATA_PATH | 데이터 파일 경로 | 내 csv 파일이 있는 자리 |
| TARGET | 맞히려는 열 | 결과 열 이름 |
| POSITIVE | 관심 갈래 | 그 열에서 찾으려는 쪽 (불합격·고장…) |
| features | 입력 열들 | 모델에 넣는 열 이름 목록 |
| DROP | 뺄 열 | 누수 점검에서 걸러낸 열 (없으면 빈 목록) |
| label | 정답표 | 찾으려는 쪽이면 1, 아니면 0 |
| X / y | 입력 / 정답 | 표 전체 입력값 / 0과 1 |
| X_train, X_test | 학습용 / 시험용 입력 | 8대 2로 나눈 것 |
| y_train, y_test | 학습용 / 시험용 정답 | 위와 짝 |
| baseline_pred | 기준 모델의 답안지 | 학습 없이 전부 0 |
| scaler | 같은 자로 맞추는 것 | StandardScaler |
| model | 학습시킨 모델 | LogisticRegression |
| y_pred | 모델이 내놓은 답 | 0과 1 |
| tp | true positive | 지목했는데 진짜였던 건수 |
| precision / recall | 정밀도 / 재현율 | 지난주에 이름 붙인 그 지표 |

| 쓴 함수 | 하는 일 |
|---|---|
| select_dtypes(include="number") | 숫자로 된 열만 골라낸다 |
| fillna(중앙값) | 빈칸을 그 열의 중앙값으로 채운다 |
| astype(int) | 참·거짓을 1과 0으로 바꾼다 |
| stratify=y | 나눌 때 비율을 양쪽에 맞춘다 |
| StandardScaler | 각 열을 같은 자로 재게 바꾼다 |
| max_iter=1000 | 계산을 끝까지 하게 한다 |

### 내 갈래

- 갈래 : A - 맞히기 (결과 열이 있다)
- 결과 열 : result / 관심 갈래 : 불합격
- 이 기준으로 정한 이유 : 현장에서 걸러내려는 게 불합격 셀이라서

In [ ]:
import pandas as pd
import numpy as np

# 여기 세 줄만 내 것으로 바꾼다
DATA_PATH = "../../data/day04_battery.csv"   # 계획표 1번 줄에 적어둔 경로
TARGET    = "result"                     # 결과 열 이름 (target = 맞히려는 것)
POSITIVE  = "불합격"                      # 그 열에서 찾으려는 쪽 (positive = 관심 갈래)

df = pd.read_csv(DATA_PATH)

# 숫자로 된 열만 입력으로 쓴다. 결과 열은 답이므로 입력에서 뺀다
# features = 모델에 넣는 입력 열들
features = [c for c in df.select_dtypes(include="number").columns if c != TARGET]

# DROP - 도전(데이터 누수 점검)에서 찾은 빼야 할 열. 없으면 [] 그대로 둡니다
DROP = []
features = [c for c in features if c not in DROP]

# 빈칸을 그 열의 중앙값으로 채운다
df[features] = df[features].fillna(df[features].median())

# 판정을 숫자로 - 찾으려는 쪽이면 1, 아니면 0 (label = 정답표)
df["label"] = (df[TARGET] == POSITIVE).astype(int)

X = df[features]
y = df["label"]

print("표:", df.shape, "/ 입력으로 쓸 열:", len(features), "개")
print("찾으려는 쪽:", int(y.sum()), "건 (", round(y.mean() * 100, 2), "% )")

표: (2847, 15) / 입력으로 쓸 열: 8 개
찾으려는 쪽: 133 건 ( 4.67 % )


In [ ]:
from sklearn.model_selection import train_test_split

# 8대 2로 나누되, 비율을 양쪽에 맞춰서
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.2,
    random_state=42,
    stratify=y                 # 비율을 양쪽에 맞추라는 뜻
)

# baseline = 기준 모델. 학습하지 않고 전부 0(정상)이라 답하는 답안지
baseline_pred = np.zeros(len(y_test), dtype=int)
baseline_accuracy = (baseline_pred == y_test).mean()

print("학습용:", len(X_train), "건 (찾으려는 쪽", int(y_train.sum()), "건 )")
print("시험용:", len(X_test), "건 (찾으려는 쪽", int(y_test.sum()), "건 )")
print("기준 모델 정확도:", round(baseline_accuracy * 100, 2), "%")

학습용: 2277 건 (찾으려는 쪽 106 건 )
시험용: 570 건 (찾으려는 쪽 27 건 )
기준 모델 정확도: 95.26 %


In [ ]:
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression

# scaler = 열마다 다른 단위를 같은 자로 맞춰주는 것
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)   # 기준은 학습용에서만 배운다
X_test_scaled  = scaler.transform(X_test)        # 시험용은 배운 기준을 적용만 한다

model = LogisticRegression(max_iter=1000)
model.fit(X_train_scaled, y_train)
y_pred = model.predict(X_test_scaled)            # y_pred = 모델이 내놓은 답

accuracy  = (y_pred == y_test).mean()                          # 정확도
predicted = int((y_pred == 1).sum())                           # 지목한 건수
tp        = int(((y_pred == 1) & (y_test == 1)).sum())         # true positive - 지목했는데 진짜였던 것
precision = tp / predicted if predicted else 0                 # 정밀도
recall    = tp / int(y_test.sum())                             # 재현율

print("기준 모델 정확도:", round(baseline_accuracy * 100, 2), "%")
print("내 모델 정확도  :", round(accuracy * 100, 2), "%")
print("지목한 건수     :", predicted, "건")
print("그중 진짜       :", tp, "건")
print("시험용의 실제   :", int(y_test.sum()), "건")
print("정밀도 precision:", round(precision, 3))
print("재현율 recall   :", round(recall, 3))

기준 모델 정확도: 95.26 %
내 모델 정확도  : 97.02 %
지목한 건수     : 14 건
그중 진짜       : 12 건
시험용의 실제   : 27 건
정밀도 precision: 0.857
재현율 recall   : 0.444


### 받은 코드 확인

| # | 확인한 것 | 어땠나 |
|---|---|---|
| 1 | 비율 맞춰 나누기 (stratify) | 들어 있었다 |
| 2 | 표준화를 학습용에서만 (fit_transform / transform) | 시험용에도 fit 이 되어 있어 고쳐달라고 했다 |
| 3 | 기준 모델 | 없어서 넣어달라고 했다. 95.26%가 여기서 나왔다 |
| 4 | 지목 건수와 재현율 | 정확도만 있어서 추가로 요청했다 |

- 고친 것이 둘 : 시험용에 fit 하지 말라고 한 것, 기준 모델을 넣어달라고 한 것
- 둘 다 안 고쳤으면 97%만 보고 잘된 줄 알았을 것이다

---

### 도전 - 하이퍼파라미터를 바꿔본 결과

In [ ]:
# class_weight - 드문 쪽을 무겁게 세라는 하이퍼파라미터
weighted_model = LogisticRegression(max_iter=1000, class_weight="balanced")
weighted_model.fit(X_train_scaled, y_train)
y_pred_weighted = weighted_model.predict(X_test_scaled)

tp_w = int(((y_pred_weighted == 1) & (y_test == 1)).sum())
print("[손 안 댐] 지목", int((y_pred == 1).sum()), "건 / 그중 진짜", tp, "건 / 재현율", round(recall, 3))
print("[가중치 ] 지목", int((y_pred_weighted == 1).sum()), "건 / 그중 진짜", tp_w, "건 / 재현율", round(tp_w / int(y_test.sum()), 3))

[손 안 댐] 지목 14 건 / 그중 진짜 12 건 / 재현율 0.444
[가중치 ] 지목 101 건 / 그중 진짜 25 건 / 재현율 0.926


| | 지목 | 그중 진짜 | 재현율 |
|---|---|---|---|
| 손 안 댐 | 14 | 12 | 0.444 |
| class_weight="balanced" | 101 | 25 | 0.926 |

- 알게 된 것 : 재현율이 0.444에서 0.926으로 올랐지만 헛걸음이 2건에서 76건으로 늘었다. 어느 쪽이 나은지는 재검사 한 건에 드는 비용을 알아야 정해진다

# 실습 19 - 결과 한 장

## v1 결과

| | 정확도 | 지목 | 그중 진짜 | 정밀도 | 재현율 |
|---|---|---|---|---|---|
| 기준 모델 (전부 합격) | 95.26% | 0건 | 0건 | - | 0.000 |
| 내 모델 v1 | 97.02% | 14건 | 12건 | 0.857 | 0.444 |

- 시험용에 실제로 있던 불합격 : 27건
- 한 줄 요약 : 정확도는 기준보다 1.76%p 올랐고 지목한 14건 중 12건이 실제 불합격이었다(정밀도 0.857). 다만 실제 27건 중 15건은 못 잡았다(재현율 0.444)

### 다음에 할 일

1. class_weight로 드문 쪽을 무겁게 세어 재현율을 올려보기 (실습 18 도전에서 한 번 돌려봤다 - 헛걸음이 크게 늘었다)
2. 재현율이 오를 때 헛걸음이 얼마나 느는지 같이 재기
3. 재검사 한 건에 드는 비용을 알아보고, 어느 쪽이 나은지 정하기

### 카드에 옮겨 적을 것

**3번 칸 - 사용 데이터**
셀 검사 기록 / 2847행 x 14열 / 결과 열은 합격·불합격 / 불합격 133건 (4.67%)

**5번 칸 - 결과 수치**
전부 합격이라 답하는 기준 모델 95.26% 대비 97.02%.
지목한 14건 중 12건이 실제 불합격이었으나(정밀도 0.857),
실제 27건 중 15건은 못 잡았다(재현율 0.444). 그래서 정확도만으로는 판단할 수 없는 문제였다

---

### 도전 - 그래서 무엇을 하자는 것인가

- 측정값이 이 조건에 걸리는 셀은 최종 검사 전에 정밀 재검사 대상으로 따로 빼두자는 것. 다만 지금 v1로는 실제 27건 중 12건만 걸러지므로, 이 상태로 현장에 거는 것은 이르다